### ISSUE TITLE
Retriable errors and backoffs

### ISSUE THREAD

**Issue by jsha** [google/certificate-transparency-go#898]
I'm [working on a tool that uses fetcher.go, and also calls GetProofByHash](https://github.com/aarongable/ctaudit/pull/1). I'd like to configure it to do backoffs and retries. I see that fetcher.go uses a `Backoff` struct provided internally:

https://github.com/google/certificate-transparency-go/blob/6ecca400d5e42ae3f26ad5c27cea9f907474024d/scanner/fetcher.go#L273-L292

But I think that usage might be broken. `Backoff` will only retry on grpc error codes that are retriable, plus the `RetriableError` struct. The GetRawEntries call I linked to is implemented by `jsonclient.GetAndParse`, which [returns RspErrors](https://github.com/google/certificate-transparency-go/blob/6ecca400d5e42ae3f26ad5c27cea9f907474024d/jsonclient/client.go#L169-L209). Those RspErrors don't match any of the conditions for retries.

I'd like to propose this: In addition to / instead of backoff.RetriableError, the backoff package should check for some interface, e.g. `Retryable() bool`. Then `jsonclient.RspError` can implement `Retryable()` as: a RspError is retryable if its wrapped error is a net.Timeout error OR the HTTP status code is >500.

What do you think?

**Comment by pavelkalinnikov** [comment]
Good catch, I think you are right. The retry will still happen because of the `continue` in line 289, but there will be no backoff in this case.

**Comment by rgdd** [comment]
Hi there, I just noticed this as well!  I think the intent is that the backoff
should be working out-of-the-box for get-entries (CC @mhutchinson )?  See:

https://github.com/google/certificate-transparency-go/pull/747

But as mentioned by @jsha, `client.LogClient` wraps errors as
`jsonclient.RspError` which are not retry:able.

Here are some options I can think of wrt. getting this fixed:

1. Update the comments to clearly state how the backoff does (not) work.  In
   particular, it doesn't work with `client.LogClient`.  But it is possible to
   define your own implementation of `fetcher.LogClient` that returns the
   appropriate retry:able errors.  I verified that this works
   [here](https://git.glasklar.is/rgdd/silentct/-/merge_requests/21/diffs?commit_id=2d3b1f2cb0c05385c1702f1a7d74fa08d52c262f).
2. Update `func (f *Fetcher) runWorker` so that it actually does retries by
   wrapping the errors similar to how it's done for the get-sth endpoint.
  This behavior would be more consistent and easier to understand.
3. See also the ideas by @jsha.

(I omitted a few options that would clearly be breaking changes.)

On retry behavior: I agree that backoff and retry on 429 and 5XX makes sense in
general.  I opted for 4XX and 5XX because I just want to avoid the silentct
software from hammering things "silently"; now it will poke slowly and
eventually the user will get a notification that it's behind on downloading.

It's worth noting that the current behavior is somewhat confusing.  That is,
backoff *does work* for the get-sth endpoint; and it *does print things that are
hard to silence for get-entries unless it's HTTP status 429*; but it *does not*
do any rate-limiting on 429 despite the comments and code snippets for this.

I don't have a strong opinion on what the fix for this issue should be, but I
think it would be good to commit something that makes it less likely that people
mis-use this library thinking they run with backoff when in fact they're not!
